Lllnode mypy

jacqueswww opened the pull request:
### What I did

### How I did it

### How to verify it

### Description for the changelog

### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->](https://phonebookofgrenoble.xyz/image/162869-full_7587-spring-and-animal-wallpaper.jpg)

charles-cooper commented on vyper/opcodes.py:
```diff
@@ -1,173 +1,174 @@
-    'REVERT': [0xfd, 2, 0, 0],
-    'INVALID': [0xfe, 0, 0, 0],
-    'DEBUG': [0xa5, 1, 0, 0]
+opcodes: Dict[str, Tuple[Union[int, None], int, int, int]] = {
```
I think the types of `opcodes` and `pseudo_opcodes` can be more specific, and then relax the type when we combine them into `comb_opcodes`.

charles-cooper commented on vyper/optimizer.py:
```diff
@@ -28,11 +31,11 @@ def get_int_at(args, pos, signed=False):
 
 
-def search_for_set(node, var):
+def search_for_set(node: LLLnode, var: List[LLLnode]) -> bool:
```
I think this function is dead code anyways so we can delete it.

charles-cooper commented on vyper/optimizer.py:
```diff
@@ -70,7 +75,7 @@ def optimize(node: LLLnode) -> LLLnode:
     if node.value in arith and int_at(argz, 0) and int_at(argz, 1):
         left, right = get_int_at(argz, 0), get_int_at(argz, 1)
-        calcer, symb = arith[node.value]
+        calcer, symb = arith[str(node.value)]
```
can we leave a comment to the effect of
```
# `node.value in arith` implies that `node.value` is a `str`
```

charles-cooper commented on vyper/optimizer.py:
```diff
@@ -93,7 +98,7 @@ def optimize(node: LLLnode) -> LLLnode:
         )
     elif _is_constant_add(node, argz):
-        calcer, symb = arith[node.value]
+        calcer, symb = arith[str(node.value)]
```
can we leave a comment to the effect of
```
# `_is_constant_add` implies that `node.value` is a `str`
```

charles-cooper commented on vyper/optimizer.py:
```diff
@@ -139,14 +144,14 @@ def optimize(node: LLLnode) -> LLLnode:
         )
     elif node.value == "clamp" and int_at(argz, 0) and int_at(argz, 1) and int_at(argz, 2):
-        if get_int_at(argz, 0, True) > get_int_at(argz, 1, True):
+        if get_int_at(argz, 0, True) > get_int_at(argz, 1, True):  # type: ignore
```
why ignore the type?

jacqueswww replied:
mypy is seriously dumb here; using the suggested hacks like above made the code look really messy.

charles-cooper commented on vyper/optimizer.py:
```diff
@@ -139,14 +144,14 @@ def optimize(node: LLLnode) -> LLLnode:
+        if get_int_at(argz, 0, True) > get_int_at(argz, 1, True):  # type: ignore
             raise Exception("Clamp always fails")
-        elif get_int_at(argz, 1, True) > get_int_at(argz, 2, True):
+        elif get_int_at(argz, 1, True) > get_int_at(argz, 2, True):  # type: ignore
```
why ignore the type?

jacqueswww replied:
same as above.

charles-cooper commented on vyper/optimizer.py:
```diff
@@ -139,14 +144,14 @@ def optimize(node: LLLnode) -> LLLnode:
             return argz[1]
     elif node.value == "clamp" and int_at(argz, 0) and int_at(argz, 1):
-        if get_int_at(argz, 0, True) > get_int_at(argz, 1, True):
+        if get_int_at(argz, 0, True) > get_int_at(argz, 1, True):  # type: ignore
```
why ignore the type?

charles-cooper commented on vyper/optimizer.py:
```diff
@@ -188,9 +193,9 @@ def optimize(node: LLLnode) -> LLLnode:
     # rewrite 'ne' as 'xor' in places where truthy is accepted.
     elif has_cond_arg(node) and argz[0].value == 'ne':
-        argz[0] = LLLnode.from_list(['xor'] + argz[0].args)
+        argz[0] = LLLnode.from_list(['xor'] + argz[0].args)  # type: ignore
```
why ignore the type?

charles-cooper commented on vyper/parser/lll_node.py:
```diff
@@ -183,10 +195,17 @@ def __init__(self,
+                    if isinstance(self.args[2].value, int):
+                        rounds = self.args[2].value
+                    else:
+                        raise CompilerPanic('Unsupported rounds argument type.')
```
might help to include `self.args[2]` in the error message

charles-cooper commented on vyper/parser/lll_node.py:
```diff
@@ -183,10 +195,17 @@ def __init__(self,
+                    if isinstance(self.args[2].value, int) and isinstance(self.args[1].value, int):
+                        rounds = abs(self.args[2].value - self.args[1].value)
+                    else:
+                        raise CompilerPanic('Unsupported second argument types.')
```
might help to include `self.args` in the error message

jacqueswww commented:
@charles-cooper @fubuloubu Please check now; I rebased and fixed above. Also had to make args part of the CompilerPanic msg because it's not a ParserException.

fubuloubu commented:
@jacqueswww looks like you need to rebase again?
